File: test-suite-kotlin/src/test/kotlin/io/micronaut/docs/config/builder/SparkPlug.kt
/*
 * Copyright 2017-2020 original authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.micronaut.docs.config.builder

internal data class SparkPlug(
        val name: String?,
        val type: String?,
        val companyName: String?
) {
    override fun toString(): String {
        return "${type ?: ""}(${companyName ?: ""} ${name ?: ""})"
    }

    companion object {
        fun builder(): Builder {
            return Builder()
        }
    }

    data class Builder(
            var name: String? = "4504 PK20TT",
            var type: String? = "Platinum TT",
            var companyName: String? = "Denso"
    ) {
        fun withName(name: String?): Builder {
            this.name = name
            return this
        }

        fun withType(type: String?): Builder {
            this.type = type
            return this
        }

        fun withCompany(companyName: String?): Builder {
            this.companyName = companyName
            return this
        }

        fun build(): SparkPlug {
            return SparkPlug(name, type, companyName)
        }
    }

}

File: test-suite-python/src/test/python/micronaut/docs/config/builder/SparkPlug.py
class SparkPlug:
    def __init__(
        self,
        name: str | None,
        type: str | None,
        company_name: str | None,
    ):
        self.name = name
        self.type = type
        self.company_name = company_name

    def get_name(self) -> str | None:
        return self.name

    def get_type(self) -> str | None:
        return self.type

    def get_company_name(self) -> str | None:
        return self.company_name

    @staticmethod
    def builder() -> "SparkPlugBuilder":
        return SparkPlugBuilder()

    def __str__(self) -> str:
        return f"{self.type or ''}({self.company_name or ''} {self.name or ''})"


class SparkPlugBuilder:
    name: str | None = "4504 PK20TT"
    type: str | None = "Platinum TT"
    company_name: str | None = "Denso"

    def withName(self, name: str | None) -> "SparkPlugBuilder":
        self.name = name
        return self

    def withType(self, type: str | None) -> "SparkPlugBuilder":
        self.type = type
        return self

    def withCompanyName(self, company_name: str | None) -> "SparkPlugBuilder":
        self.company_name = company_name
        return self

    def build(self) -> SparkPlug:
        return SparkPlug(self.name, self.type, self.company_name)


Builder = SparkPlugBuilder
SparkPlug.Builder = SparkPlugBuilder
